From 728a7f2e60fb1034c5cf4096522306f2f30424ff Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Thu, 15 Dec 2016 21:54:52 +0800
Subject: [PATCH 1439/2241] mmc: mvebu: add PM support for XENON MMC driver

- In order to support low power mode, the XENON driver needs
  to support suspend and resume.
- Enable clk first and initialize Xenon again during
  mmc resume action.

Change-Id: I501816f4753ccfdb3d06d5b5f57839e808986424
Signed-off-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/34682
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
Signed-off-by: zachary <zhangzg@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37169
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Kostya Porotchkin <kostap@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 drivers/mmc/host/sdhci-xenon.c | 46 +++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 45 insertions(+), 1 deletion(-)

diff --git a/drivers/mmc/host/sdhci-xenon.c b/drivers/mmc/host/sdhci-xenon.c
index 0b693fa..a770a7e 100644
--- a/drivers/mmc/host/sdhci-xenon.c
+++ b/drivers/mmc/host/sdhci-xenon.c
@@ -591,6 +591,48 @@ static int sdhci_xenon_remove(struct platform_device *pdev)
 	return 0;
 }
 
+#ifdef CONFIG_PM
+static int sdhci_xenon_suspend(struct device *dev)
+{
+	int ret;
+	struct sdhci_host *host = dev_get_drvdata(dev);
+	struct sdhci_pltfm_host *pltfm_host = sdhci_priv(host);
+
+	ret = sdhci_suspend_host(host);
+
+	if (pltfm_host->clk)
+		clk_disable_unprepare(pltfm_host->clk);
+
+	return ret;
+}
+
+static int sdhci_xenon_resume(struct device *dev)
+{
+	int ret;
+	struct sdhci_host *host = dev_get_drvdata(dev);
+	struct sdhci_pltfm_host *pltfm_host = sdhci_priv(host);
+
+	if (pltfm_host->clk)
+		ret = clk_prepare_enable(pltfm_host->clk);
+
+	ret = xenon_sdhc_probe(host);
+
+	/* Initialize SoC PAD register for MMC PHY voltage
+	 * For eMMC, it is set to 1.8V
+	 * For SD/SDIO, it is set to 3.3V
+	 */
+	xenon_soc_pad_ctrl(host, MMC_SIGNAL_VOLTAGE_330);
+
+	ret = sdhci_resume_host(host);
+
+	return ret;
+}
+
+static const struct dev_pm_ops sdhci_xenon_pmops = {
+	SET_SYSTEM_SLEEP_PM_OPS(sdhci_xenon_suspend, sdhci_xenon_resume)
+};
+#endif
+
 static const struct of_device_id sdhci_xenon_dt_ids[] = {
 	{ .compatible = "marvell,armada8k-sdhci",},
 	{ .compatible = "marvell,armada-3700-sdhci",},
@@ -602,7 +644,9 @@ static struct platform_driver sdhci_xenon_driver = {
 	.driver	= {
 		.name	= "xenon-sdhci",
 		.of_match_table = sdhci_xenon_dt_ids,
-		.pm = &sdhci_pltfm_pmops,
+#ifdef CONFIG_PM
+		.pm = &sdhci_xenon_pmops,
+#endif
 	},
 	.probe	= sdhci_xenon_probe,
 	.remove	= sdhci_xenon_remove,
-- 
2.7.4

